### ISSUE TITLE
Using <LinkTo> dynamic @route inside engines does not work as expected

### ISSUE THREAD

**Issue by abhilashlr** [ember-engines/ember-engines#692]
I've an Ember app (3.16) containing an in-repo addon "ui" and an in-repo engine "admin".
Both the app and the engine share this addon.
One of the components of the addon contains a dynamic `<LinkTo>` usage like below:
```hbs
{{#each @menuItems as |menuItem|}}
  <div class="main-navbar-item">
    {{#if menuItem.model}}
      <LinkTo @route={{menuItem.route}} @model={{menuItem.model}}>
        {{t menuItem.textKey}}
      </LinkTo>
    {{else}}
      {{!-- Tried this approach as well
      {{#link-to menuItem.route}}
        {{t menuItem.textKey}}
      {{/link-to}} --}}
      <LinkTo @route={{menuItem.route}}>
        {{t menuItem.textKey}}
      </LinkTo>
    {{/if}}
  </div>
{{/each}}
```
Using this component inside the host app renders the links well and as expected. While trying to use the same component in the engine and rendering the same, throws the following error:

```
Error: Assertion Failed: You attempted to update `route` on `(unknown object)`, but it had already been used previously in the same computation.  Attempting to update a value after using it in a computation can cause logical errors, infinite revalidation bugs, and performance issues, and is not supported.

`route` was first used:

- While rendering:
  ----------------
    unknownObject.route
...
```

Am I missing something or is this a bug?

**Comment by roncodes** [comment]
Also experiencing the same problem on Octane

**Comment by roncodes** [comment]
I made a pull request, but have a fix in my fork to use in meantime

**Comment by sbflynn** [comment]
Also experiencing this in my engines that use a standalone (ie not in-repo) addon containing a <LinkTo/> element with a dynamic route - blocks our road to Octane.

**Comment by abhilashlr** [comment]
@buschtoens @villander can you please take a look at this issue and the PR made?

**Comment by chriskrycho** [comment]
This dates back to at least 3.12. Here's the outline of the problem as I discovered it:

1. Engines prefix route names by their mount point: [here](https://github.com/ember-engines/ember-engines/blob/d808e6b505f215c32028f210393144468bf8d2d3/addon/components/link-to-component.js#L23).
2. This is applied by calling `set` on the `LinkTo` component: [here](https://github.com/ember-engines/ember-engines/blob/d808e6b505f215c32028f210393144468bf8d2d3/addon/components/link-to-component.js#L52).
3. Calling `set` triggers `notifyPropertyChange` (this is how CPs get updated properly in general, not specific to engines).
4. `notifyPropertyChange` attempts to trigger an update on the property on the backing class for the component which invoked `link-to`, which is a CP marked as `.readOnly()`.
5. :boom:

The issue found in the original is a variant on the same thing. It's using the tracked properties tags-based update instead, but the problem is the same underlying issue: in that case it's updating (via `set`) a value that has already been "pulled" during the update. ([This post](https://www.pzuraq.com/how-autotracking-works/) provides background for this, and [this section in particular](https://www.pzuraq.com/how-autotracking-works/#principle1predictableoutput) walks through why this assertion exists with autotracking.)

**Comment by chriskrycho** [comment]
Further investigation: this is due to the changes made to the implementation of `<LinkTo>`  in 3.10, and specifically because the previous implementation stored the target route in `targetRouteName`, where the new one uses `route`, and `route` is a value on the `args` to `LinkTo`, so [this lookup](https://github.com/emberjs/ember.js/blob/d96d9aad52e010da977d813f479494397ee53ec6/packages/%40ember/-internals/glimmer/lib/component.ts#L787) finds a reference to the property now when it didn't before.

**Comment by chriskrycho** [comment]
cc. @chancancode @pzuraq @rwjblue

**Comment by amk221** [comment]
Can anyone confirm if this is the same issue: [small reproduction repo](https://github.com/amk221/-ember-engine-link-to-current)

**Comment by amk221** [comment]
Think it is. Fixing for now with `@route={{unbound this.myRoute}}`

**Comment by levimllr** [comment]
Echoing what others have said, this issue is very real. Thanks @chriskrycho for outlining the problem! I racked my brain on this guy for a good spell and ultimately reproduced it in [a very simple Ember app + engine](https://github.com/levimllr/ember-engine-router-weirdness).

I've worked around it with `@route="{{this.myRoute}}"`, but @amk221 I like your way better 👍

**Comment by malatin3** [comment]
Has there been any movement on this issue? Been blocking us from upgrading since early in the year.

**Comment by chancancode** [comment]
Oh huh. I had no idea that the engine addon reopened the link component (which we [just deprecated](https://github.com/emberjs/rfcs/pull/671)). Does anyone know if this separation is necessary at all? (i.e. what functionality does the engine addon provide on top of the basic engine support in ember, that requires the additional link-to functionality to be separated out?)

**Comment by rwjblue** [comment]
I believe that this was resolved by #735 (released in 0.8.8). I think in the long run that we will resolve this category of issues much better by moving core functionality for engines (like the link-to modifications, and possibly link-to-external) into Ember (as @chancancode is looking into).

**Comment by chancancode** [comment]
I submitted a [PR to ember](https://github.com/emberjs/ember.js/pull/19223) to fix this, can I get a review from people who are more familiar with engines?
